kore/src/Kore/Step/BaseStep.hs

```diff
@@ -890,3 +863,526 @@ unwrapPredicateVariables
+not applied to the rule (see 'applyRule').
+
+ -}
+unifyRule
+    ::  ( Ord     (variable Object)
+        , Show    (variable Object)
+        , Unparse (variable Object)
+        , FreshVariable  variable
+        , SortedVariable variable
+        )
+    => MetadataTools Object StepperAttributes
+    -> UnificationProcedure Object
+    -> PredicateSubstitutionSimplifier Object
+    -> StepPatternSimplifier Object
+    -> BuiltinAndAxiomSimplifierMap Object
+
+    -> ExpandedPattern Object variable
+    -- ^ Initial configuration
+    -> RulePattern Object variable
+    -- ^ Rule
+    -> BranchT
+        (ExceptT (StepError Object variable) Simplifier)
+        (UnifiedRule variable)
+unifyRule
+    metadataTools
+    (UnificationProcedure unificationProcedure)
+    predicateSimplifier
+    patternSimplifier
+    axiomSimplifiers
+
+    initial@Predicated { term = initialTerm }
+    rule
+  = do
+    let
+        -- Rename free axiom variables to avoid free variables from the initial
+        -- configuration.
+        (_, rule') = RulePattern.refreshRulePattern configVariables rule
+        configVariables = ExpandedPattern.freeVariables initial
+        RulePattern { left = ruleLeft } = rule'
+    unifier <- unifyPatterns ruleLeft initialTerm
+    return (unifier $> rule')
+  where
+    unifyPatterns pat1 pat2 = do
+        (unifiers, _) <-
+            liftFromUnification
+            $ Monad.Trans.lift
+            $ unificationProcedure
+                metadataTools
+                predicateSimplifier
+                patternSimplifier
+                axiomSimplifiers
+                pat1
+                pat2
+        scatter unifiers
+
+{- | Instantiate the rule by applying the unification solution.
+
+The unification solution is normalized with the 'requires' clause from the
+unified rule. @instantiateRule@ fails if normalization fails. @instantiateRule@
+branches when the 'PredicateSubstitutionSimplifier' causes normalization to
+branch.
+
+ -}
+instantiateRule
+    ::  ( Ord     (variable Object)
+        , Show    (variable Object)
+        , Unparse (variable Object)
+        , FreshVariable  variable
+        , SortedVariable variable
+        )
+    => MetadataTools Object StepperAttributes
+    -> PredicateSubstitutionSimplifier Object
+    -> StepPatternSimplifier Object
+    -> BuiltinAndAxiomSimplifierMap Object
+
+    -> UnifiedRule variable
+    -- ^ Rule and unification solution
+    -> BranchT
+        (ExceptT (StepError Object variable) Simplifier)
+        (UnifiedRule variable)
+instantiateRule
+    metadataTools
+    predicateSimplifier
+    patternSimplifier
+    axiomSimplifiers
+
+    unified@Predicated { term = axiom }
+  = do
+    let unifier = unified { term = () }
+        RulePattern { requires, ensures } = axiom
+        merged =
+            PredicateSubstitution.fromPredicate requires
+            *> PredicateSubstitution.fromPredicate ensures
```

Comment:
> Why is it ok to drop the term with `<*` and `*>` without knowing that it's `()`?

We drop terms all the time; for example, when we apply a rule we drop the initial term and replace it with the right-hand side of the rule.

> The Applicative instance lets us do that, so we must be very careful with the types we're handling (i.e. if we ever change a function to return an `ExpandedPattern` instead of a `PredicateSubstitution`, we would have to check all places where its result might be used with, say, `*>`).

This is a problem with or without using `Applicative`. There are numerous places—even without counting the uses of `*>`—where we assume we have been returned a `PredicateSubstitution` and drop the `term`.

> Wouldn't the code be more readable and safe if we were using named functions to do some of these operations?

Are you arguing against using `Applicative` for `Predicated` everywhere, or only here? I will only address this specific use: I do not think it would be more readable. The ways this usage could be unsafe are: 1. we change the return type of `PredicateSubstitution.fromPredicate` (that would be deliberately evil and I expect we will not do that) or 2. we change the type of `instantiateRule` to be more permissive (in which case we would need to inspect the body of the function to see if it can operate over the more permissive type—which is true no matter if we use `Applicative` or less permissively-typed functions).

Reply:
I'm saying that it's easy to drop a term by mistake, so, as a rule, we should have something explicit for dropping them. We used `term=()` in several places and that's sort of fine because it's explicit. It could be better if we would return something without the term, or we would check that we can drop the term (i.e. the entire thing is top, bottom, or the term is top) and crash otherwise.

`*>` makes it very easy to implicitly drop the term. That's only a small part of `Applicative`, I'm not against using `<*>` since that applies a function which, supposedly, wants to handle things that have terms.